### ISSUE TITLE
style-guide, code and naming conventions

### ISSUE THREAD

**Issue by tunnckoCore** [regexhq/regexhq#6]
@regexps/owners 
Starts here https://github.com/regexps/regex-block-comment/issues/1 

@jonschlinkert apologize in advance if something, but I think here is the correct place for it. I cant realize and understand, yet, why you keep/started it there. lol
IMO, we should keep these type of discussions here 1) because this is the main repo for this purpose, 2) history/log about what happens. Discussion shouldn't be just in some repo that will not exists in future **or** will be with some totally different target - e.g. block comment regex talks. Yea maybe closed issue, but there. Or worst - deleted repo - we lost history/votes/thoughts.
## Style-guide & Code conventions

imo, we definitely should have separate repo for style-guide
1. Should we have style-guide repo? Or just follow some other guide(s)?
   - [felixge/node-style-guide](https://github.com/felixge/node-style-guide)
   - [jshttp/style-guide](https://github.com/jshttp/style-guide) 
   - [jscs presets](https://github.com/jscs-dev/node-jscs#presets)
2. Should we have template files for repos as in `jshttp`? Which to be the files?
3. JSHint or JSLint? imo, JSHint + @jscs-dev looks great. 
4. I think we should have this minimum of files
   - package.json
   - index.js
   - readme.md
   - history.md
   - license.md - dot md, prettier look? lol [like here](https://github.com/tunnckoCore/koa-better-body/blob/master/license.md)
   - .jshintrc
   - .jscsrc
   - .editorconfig
#### Is this `.editorconfig` enough for start?

```
root = true

[*.js]
indent_size = 2
indent_style = space
charset = utf-8
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = false

[*.php]
# ...

[*.go]
# ...
```
## Naming conventions

What i think so far
- we should keep current organization name - `regexps` 
  - also suggested `regular-expressions` and `regexes`
- we should use prefix `regexp-*` for future modules, imo; or
  - prefix: regex-foo-bar, regexp-foo-bar
  - suffix: foo-bar-regex, foo-bar-regexp
  - or without worry of what is the name
- publish with prefix in npm
- without prefix in github repo
### Vote and discuss

> This is my and suggested thoughts.
> Not attack me for that I only copy-pasting everything around. I just trying to centralize the things.

_Good night from me :)_

**Comment by jonschlinkert** [comment]
I don't have any opinions on any of this. I just want good quality regex projects to use, for users to be able to find them easily, and for them all to be consistent.

**Comment by tunnckoCore** [comment]
Agreed.

**Comment by tomByrer** [comment]
> I just want good quality regex projects to use

Same here.  I'm indifferent... OTOH, boy that is alot of overhead for [1 line](https://github.com/regexps/regex-email/blob/master/index.js#L10) of [code](https://github.com/regexps/regex-utc-date/blob/master/index.js#L8).  I'm assuming you're only targeting Node?

I'm starting to think there should be only 1 repo, with each rule in its own folder.... or some way to sync the support files like `.jshintrc`, `.jscsrc`, `.editorconfig`.

**Comment by jonschlinkert** [comment]
> alot of overhead

What overhead?

**Comment by johno** [comment]
I think they will be best in their own repos. This will keep things modular and lightweight.

The overhead/boilerplate can be handled with a yeoman generator (or some other alternative). Once we have an established pattern for the regexes, I'd imagine that we will tackle some sort of generator to keep the workflow quick and simple.

Ideally only requiring the following steps:
1. `yo regex-package regex-name`
2. Add regex to index.js
3. Add regex details to the readme
4. Push to Github

Relevant issue: https://github.com/regexps/meta/issues/5

**Comment by tomByrer** [comment]
> The overhead/boilerplate can be handled with a yeoman generator

Good idea, hopefully the Gulp version.
If it is only a few LOC, maybe an online repo-maker would work?

**Comment by tunnckoCore** [comment]
> alot of overhead

Yea you have the right for 1-line modules. But as @johnotander said it can be handled with generator. My flow is fully through cli - creating gh repo, init repo, populating template files with data, generating tests and etc.

**Comment by yoshuawuyts** [comment]
There's little overhead for having modules, especially if you specify a `files` property in `package.json` ([example](https://github.com/regexps/regex-email/blob/master/package.json#L23-L27)).

If you're looking for a module generator, both `regex-utc-date` and `regex-email` were generated with [initialize](https://github.com/yoshuawuyts/initialize). Depending on your needs / taste it might be useful.

I'm -1 on a org level style guide. The only thing I think we need to agree on is how we want to expose the `regexes`, some of the discussion is already taking place [here](https://github.com/regexps/regex-utc-date/pull/1).

**Comment by jonschlinkert** [comment]
Oh, I see. Sorry my question wasn't meant to be flippant when I asked "what overhead?". @tomByrer  you meant all of the junk that comes with a repo for just a small regex. got it. I think we're all on the same page there.

I'm also -1 on an org level style guide. I think we'll have good stewardship here, and everyone in the org is experienced and opinionated about how they like to do things - which is good and expected.

just my 2c, I'm good with how things are progressing

**Comment by yocontra** [comment]
I don't think we need any of this. Maybe we should focus more on building regular expressions and less on guidelines.

![](http://www.waltongardenproducts.ie/Graphics/Dynamic/Cache/Products/1098-633918926907187500_wgb_cm-scale_w-800_h-710_q-80_wgb.jpg)

Let best practices emerge naturally.

**Comment by tunnckoCore** [comment]
So okey, Im agree with you @regexps/owners , guys. Let's finalize the things?
- Organization name remains the same.
- No specific style guide.
  - Only `.editorconfig`! Indent size 2, indent style spaces, please, lol
- keyword `regex` as suffix or prefix, no matter, but requirement
  - because search show more results
  - sounds more natural
  - commonly used

Something missing?

**Comment by johno** [comment]
:+1:

**Comment by jonschlinkert** [comment]
:+1:

**Comment by tunnckoCore** [comment]
Okey you can transfer your repos/projects :P

What about this badge [![regexps org](http://img.shields.io/badge/regexps-approved-brightgreen.svg)](https://github.com/regexps) as requirement? lol

**Comment by arthurvr** [comment]
I'm going to take the time to add editorconfigs to all projects, everybody okay with that? I'd go for this one, as a start:

```
# For more information about the properties used in
# this file, please see the EditorConfig documentation:
# http://editorconfig.org/

root = true

[*]
charset = utf-8
end_of_line = lf
indent_size = 2
indent_style = space
insert_final_newline = true
trim_trailing_whitespace = true

[*.md]
trim_trailing_whitespace = false
```

On the while I could make sure all repos follow this.

**Comment by arthurvr** [comment]
Also we should standardise readme files, something like:

```markdown
# title

<optional> badges for npm, tests, ... </optional>

> Short description of the regex.

<optional>Created by [John Doe](http://johndoe.com)</optional>
<optional>Maintained by [John Doe](http://johndoe.com)</optional>
<optional other acknowledgements />

## Installation
'''sh
npm install example-regex
'''

## Usage
'''js
var example = require('example-regex');
// examples showing how to use the regex
'''

## License

This regexp is licensed under [MIT](/license).

<optional>
## Todo
- [ ] Todo 1
- [ ] Todo 2
- [ ] ...
</optional>

```

**Comment by arthurvr** [comment]
If we can agree on the readme formatting, I could make sure they're standardised between projects. Also, I can wrap up this hole discussion by a page in the readme, if you guys think that's useful.

**Comment by yoshuawuyts** [comment]
> I'm going to take the time to add editorconfigs to all projects

Cool, PRs are welcome!

> Also we should standardise readme files

@arthurvr -1, [consensus was already reached](https://github.com/regexps/meta/issues/6#issuecomment-66709756). A better investment of time would probably be to create more modules.

**Comment by tunnckoCore** [comment]
> I'm going to take the time to add editorconfigs to all projects

@arthurvr awesome, cool :+1:

**Comment by tomByrer** [comment]
So use a `var` or simply `return`?  EG
[`return /(?:\/|^)(\.[^\/]*)$/;`](https://github.com/regexps/dotfile-regex/blob/master/index.js#L11)
vs
[`var regex = /(?:youtube\.com\/\...)/g; return regex;`](https://github.com/regexps/youtube-regex/blob/master/index.js#L11-L13)

**Comment by tunnckoCore** [comment]
@tomByrer no matter. its "so so", lol

**Comment by tomByrer** [comment]
Cosi Cosi?
Well, I was thinking of an automated scrapper.  Better for something standardized, then have to make a complicated RegEx to scrape RegExes.

**Comment by tunnckoCore** [comment]
@regexps/owners 
I think we can close this and summarize it when generators #5 / scaffolders and website is ready?

For the start:
- future packages to include the keyword, no matter as prefix or suffix
- already existing packages to force code and readme style
- in javascript context, package to be function that return the regex
- what for other languages?

I have a lot more Qs, but.. hm. Maybe I'll open new thread "General discussion", cuz they cover many related between them topcis and steps that we should do.
